### DISCUSSION TITLE
Better way to limit property mapping in object within other object

### DISCUSSION THREAD

**Discussion by Kartik-Boreda-Imprivata** [mapstruct/mapstruct#3416] [Q&A] [answered]
I have a scenario where there is an `OrderItem` entity with an `Order` entity in it, and I'm trying to map only the id property from `Order` to the id property in `OrderDTO.` 
Is there a way other than using `BeanMapping#ignoreMapping` because I must map all other properties manually in `OrderItem` to DTO?

I tried `@Mapping(target = "orderDto.id", source="order.id")`, but MapStruct maps all the properties anyway.

**Comment by thunderhook** [comment, marked as answer]
I hope I understand you correctly.

Given this mapper:
```java
class OrderItem {
    public Long id;
    public String itemName;
    public Order order;
}

class Order {
    public Long id;
    public String orderName;
}

class OrderItemDto {
    public Long id;
    public String itemName;
    public OrderDto orderDto;
}

class OrderDto {
    public Long id;
    public String orderName;
}

@Mapper
interface OrderMapper {

    @Mapping(target = "orderDto.id", source="order.id")
    OrderItemDto map(OrderItem source);

}
```

This will implicitly map anything.

When you add `BeanMapping#ignoreMapping` then the implicit mapping on the level of `OrderItem/Dto` is disabled, but not on the level of `Order/Dto`:

```java
// GENERATED CODE
    @Override
    public OrderItemDto map(OrderItem source) {
        if ( source == null ) {
            return null;
        }

        OrderItemDto orderItemDto = new OrderItemDto();

        orderItemDto.orderDto = orderToOrderDto( source.order );

        return orderItemDto;
    }

    protected OrderDto orderToOrderDto(Order order) {
        if ( order == null ) {
            return null;
        }

        OrderDto orderDto = new OrderDto();

        orderDto.id = order.id;
        orderDto.orderName = order.orderName;

        return orderDto;
    }
```

When you also want to disable the explicit mapping on the level of `Order/Dto`, then this can be achieved by un-nesting and providing an additional mapping method:

```java
@Mapper
interface OrderMapper {

    @BeanMapping(ignoreByDefault = true)
    OrderItemDto map(OrderItem source);

    @Mapping(target = "id", source="id")
    @BeanMapping(ignoreByDefault = true)
    OrderDto map(Order source);

}
```

This will generated the desired mapping:
```java
// GENERATED CODE
    @Override
    public OrderItemDto map(OrderItem source) {
        if ( source == null ) {
            return null;
        }

        OrderItemDto orderItemDto = new OrderItemDto();

        return orderItemDto;
    }

    @Override
    public OrderDto map(Order source) {
        if ( source == null ) {
            return null;
        }

        OrderDto orderDto = new OrderDto();

        orderDto.id = source.id;

        return orderDto;
    }
```

**Reply by Kartik-Boreda-Imprivata** [reply]
Thanks for the help. My solution was similar to the example you provided but with about 20 subclasses like `OrderItem`, each with a class like `Order` from which only ID is required. 
I was hoping to find if there was a better alternative so that I could avoid creating 20 mapping methods that mapped only id from the class similar to `Order`.

**Reply by filiphr** [reply]
What you can try is to use mapping config. if we assume that all DTOs are an instance of a `BaseDto` class and all non dtos are an instance of `Base`

then you could do

```java
@MapperConfig(mappingInheritanceStrategy = MappingInheritanceStrategy.AUTO_INHERIT_FROM_CONFIG)
public interface BaseConfig {

    @BeanMapping(ignoreByDefault = true)
    @Mapping(target = "id")
    BaseDto map(Base base)
}
```

and then your mapper can look like:

```java
@Mapper(config = BaseConfig.class) {

    OrderDto map(Order order);
}
```

Note: I haven't tested this, but I think it should work.
